### PR TITLE
Add TABLESAMPLE support

### LINKED ISSUE #5871: TABLESAMPLE for multi-shard queries
We currently reject TABLESAMPLE queries across shards:
```sql
select * from test tablesample system (0.5);
ERROR:  could not run distributed query which use TABLESAMPLE
HINT:  Consider using an equality filter on the distributed table's partition column.
```

### PR DESCRIPTION
**Pull request by marcocitus**
DESCRIPTION: Add TABLESAMPLE support

This PR adds support for TABLESAMPLE in multi-shard queries.

Fixes #5871

### CODE REVIEW

File: src/test/regress/expected/multi_simple_queries.out [hunk 1]
```diff
@@ -664,6 +664,62 @@ DEBUG:  query has a single distribution column value: 1
...
+ count
+---------------------------------------------------------------------
+     0
+(1 row)
+
+SELECT count(*) FROM (SELECT random() FROM articles TABLESAMPLE SYSTEM (100)) a;
+DEBUG:  Router planner cannot handle multi-shard select queries
+ count
+---------------------------------------------------------------------
+    50
```
Comment (JelteF) [inline src/test/regress/expected/multi_simple_queries.out:713]: Are there only 50 rows in the `articles` table?

File: src/test/regress/sql/multi_simple_queries.sql [hunk 2]
```diff
@@ -319,6 +319,16 @@ SELECT * FROM articles TABLESAMPLE BERNOULLI (0) WHERE author_id = 1;
...
 
+SELECT count(*) FROM articles TABLESAMPLE SYSTEM (0);
+SELECT count(*) FROM articles TABLESAMPLE BERNOULLI (0);
+SELECT count(*) FROM articles TABLESAMPLE SYSTEM (100);
+SELECT count(*) FROM articles TABLESAMPLE BERNOULLI (100);
+
+SELECT count(*) FROM (SELECT random() FROM articles TABLESAMPLE SYSTEM (0)) a;
+SELECT count(*) FROM (SELECT random() FROM articles TABLESAMPLE BERNOULLI (0)) a;
+SELECT count(*) FROM (SELECT random() FROM articles TABLESAMPLE SYSTEM (100)) a;
+SELECT count(*) FROM (SELECT random() FROM articles TABLESAMPLE BERNOULLI (100)) a;
```
Comment (JelteF) [inline src/test/regress/sql/multi_simple_queries.sql:330]: I think this feature deserves a few more tests.

1. TABLESAMPLE with DML
2. TABLESAMPLE with some other percent than 0 and 100. to make the result consistent you can use the REPEATABLE clause.

**Review by JelteF — CHANGES_REQUESTED** [review]
Always nice to see adding a feature is less code than the error messages😄. But I think the description of the PR should explain why the concern you raised in the linked issue does not apply:

>
>Single shard queries work fine, though we're only sampling over a subset of the table so there's a question of whether the probability should be adjusted since TABLESAMPLE is assumed to apply to the whole table.
> ```sql
> select * from test tablesample system (0.5) where x = 1;
> ```

**Comment by marcocitus** [comment]
> But I think the description of the PR should explain why the concern you raised in the linked issue does not apply:

Adjusted the description. We can repurpose the issue after this PR. For multi-shard queries we do the right thing by scanning the whole table. For router/pruned queries we have a bug that is not addressed by this PR.

File: src/test/regress/sql/multi_simple_queries.sql (hunk 2 above)
Comment (marcocitus) [reply src/test/regress/sql/multi_simple_queries.sql:330]: > to make the result consistent you can use the REPEATABLE clause.

consistent with what?

File: src/test/regress/expected/multi_simple_queries.out (hunk 1 above)
Comment (marcocitus) [reply src/test/regress/expected/multi_simple_queries.out:713]: > Are there only 50 rows in the articles table

yes, there are 50 numbered inserts above.

File: src/test/regress/sql/multi_simple_queries.sql (hunk 2 above)
Comment (marcocitus) [reply src/test/regress/sql/multi_simple_queries.sql:330]: > TABLESAMPLE with DML (either erroring or working)

added

Comment (marcocitus) [reply src/test/regress/sql/multi_simple_queries.sql:332]: > TABLESAMPLE with some other percent than 0 and 100. to make the result consistent you can use the REPEATABLE clause.

added

File: src/test/regress/expected/multi_simple_queries.out [hunk 3]
```diff
@@ -664,6 +664,80 @@ DEBUG:  query has a single distribution column value: 1
...
+ count
+---------------------------------------------------------------------
+     0
+(1 row)
+
+SELECT count(*) FROM articles TABLESAMPLE SYSTEM (90) REPEATABLE (200);
+DEBUG:  Router planner cannot handle multi-shard select queries
+ count
+---------------------------------------------------------------------
+    50
```
Comment (marcocitus) [inline src/test/regress/expected/multi_simple_queries.out:685]: Since everything fits on a single page, this is giving a single page from each shard.

Comment (marcocitus) [reply src/test/regress/expected/multi_simple_queries.out:688]: added a comment

**Review by JelteF — APPROVED** [review]

**Pull request merged**